File: minigit-3/ts/src/commands/commit.ts
// commit command - Create a new commit

import * as fs from 'fs';
import * as path from 'path';
import { TreeEntry, CommitObject } from '../types';
import { findRepoRoot, getAuthorInfo, getCommitterInfo, formatAuthor, shortSha, getFileMode, normalizePath } from '../utils';
import { writeTree, writeCommit, parseCommit, readObject } from '../objects';
import { readIndex, writeIndex } from '../index-file';
import { getHeadCommit, updateHead, getCurrentBranch } from '../refs';

export function commit(args: string[]): void {
  const repoRoot = findRepoRoot();
  if (!repoRoot) {
    console.error('fatal: not a git repository');
    process.exit(1);
  }

  let message = '';
  let amend = false;
  let autoStage = false;

  // Parse arguments
  for (let i = 0; i < args.length; i++) {
    const arg = args[i];
    if (arg === '-m' && i + 1 < args.length) {
      message = args[++i];
    } else if (arg === '--amend') {
      amend = true;
    } else if (arg === '-a') {
      autoStage = true;
    }
  }

  if (!message && !amend) {
    console.error('error: no commit message specified');
    process.exit(1);
  }

  // Auto-stage modified tracked files if -a
  if (autoStage) {
    autoStageModified(repoRoot);
  }

  const entries = readIndex(repoRoot);

  if (entries.length === 0) {
    console.error('nothing to commit');
    process.exit(1);
  }

  const headCommit = getHeadCommit(repoRoot);

  // Check if there are changes to commit
  if (headCommit && !amend) {
    const headObj = readObject(headCommit, repoRoot);
    const headCommitData = parseCommit(headObj.content);
    const currentTree = buildTree(entries, repoRoot);

    if (currentTree === headCommitData.tree) {
      console.error('nothing to commit, working tree clean');
      process.exit(1);
    }
  }

  // Build tree from index
  const treeSha = buildTree(entries, repoRoot);

  // Get author and committer info
  const author = getAuthorInfo();
  const committer = getCommitterInfo();

  // Build commit
  let parents: string[] = [];

  if (amend && headCommit) {
    // Get parent from previous commit
    const prevCommitObj = readObject(headCommit, repoRoot);
    const prevCommit = parseCommit(prevCommitObj.content);
    parents = prevCommit.parents;

    // Use original message if no new message provided
    if (!message) {
      message = prevCommit.message;
    }
  } else if (headCommit) {
    parents = [headCommit];
  }

  const commitObj: CommitObject = {
    tree: treeSha,
    parents,
    author: formatAuthor(author.name, author.email, author.date),
    committer: formatAuthor(committer.name, committer.email, committer.date),
    message
  };

  const commitSha = writeCommit(commitObj, repoRoot);

  // Update HEAD
  updateHead(commitSha, repoRoot);

  // Output
  const branch = getCurrentBranch(repoRoot);
  const branchDisplay = branch ? `[${branch} ${shortSha(commitSha)}]` : `[${shortSha(commitSha)}]`;
  console.log(`${branchDisplay} ${message.split('\n')[0]}`);
}

function buildTree(entries: { name: string; sha: string; mode: number }[], repoRoot: string): string {
  // Group entries by top-level directory
  const tree = new Map<string, { name: string; sha: string; mode: number }[]>();

  for (const entry of entries) {
    const parts = entry.name.split('/');
    const topLevel = parts[0];

    if (parts.length === 1) {
      // File at root level
      if (!tree.has('')) {
        tree.set('', []);
      }
      tree.get('')!.push({ name: entry.name, sha: entry.sha, mode: entry.mode });
    } else {
      // File in subdirectory
      if (!tree.has(topLevel)) {
        tree.set(topLevel, []);
      }
      tree.get(topLevel)!.push({
        name: parts.slice(1).join('/'),
        sha: entry.sha,
        mode: entry.mode
      });
    }
  }

  // Build tree entries
  const treeEntries: TreeEntry[] = [];

  // First, add root-level files
  const rootFiles = tree.get('') || [];
  for (const file of rootFiles) {
    treeEntries.push({
      mode: file.mode.toString(8).padStart(6, '0'),
      type: 'blob',
      sha: file.sha,
      name: file.name
    });
  }

  // Then, recursively build subtrees
  for (const [dir, subEntries] of tree) {
    if (dir === '') continue;

    const subtreeSha = buildTree(subEntries, repoRoot);
    treeEntries.push({
      mode: '040000',
      type: 'tree',
      sha: subtreeSha,
      name: dir
    });
  }

  return writeTree(treeEntries, repoRoot);
}

function autoStageModified(repoRoot: string): void {
  const entries = readIndex(repoRoot);
  const newEntries: typeof entries = [];

  for (const entry of entries) {
    const absPath = path.join(repoRoot, entry.name);

    if (!fs.existsSync(absPath)) {
      // File deleted - don't include in new index
      continue;
    }

    const stats = fs.statSync(absPath);

    // Check if file was modified
    if (stats.mtimeMs / 1000 !== entry.mtimeSec + entry.mtimeNsec / 1000000000) {
      // Re-read and update entry
      const { writeBlob } = require('../objects');
      const content = fs.readFileSync(absPath);
      const sha = writeBlob(content, repoRoot);
      const mode = getFileMode(absPath);

      newEntries.push({
        ...entry,
        sha,
        mode,
        mtimeSec: Math.floor(stats.mtimeMs / 1000),
        mtimeNsec: Math.floor((stats.mtimeMs % 1000) * 1000000),
        size: stats.size
      });
    } else {
      newEntries.push(entry);
    }
  }

  writeIndex(newEntries, repoRoot);
}

File: minigit-3/py/commands/commit.py
"""commit command - Create a commit."""

import sys
from pathlib import Path


def run(args: list[str]) -> int:
    """Create a new commit."""
    from utils import find_repo_root
    from objects import create_commit, build_tree_from_index, read_commit
    from index import read_index, write_index, add_to_index
    from refs import get_head_sha, update_head_for_commit

    repo_root = find_repo_root()
    if repo_root is None:
        print("error: not a minigit repository", file=sys.stderr)
        return 1

    # Parse arguments
    message = None
    all_flag = False
    amend_flag = False

    i = 0
    while i < len(args):
        arg = args[i]
        if arg == '-m' and i + 1 < len(args):
            message = args[i + 1]
            i += 2
        elif arg == '-a':
            all_flag = True
            i += 1
        elif arg == '--amend':
            amend_flag = True
            i += 1
        else:
            i += 1

    if message is None and not amend_flag:
        print("error: must provide commit message with -m", file=sys.stderr)
        return 1

    # Handle -a flag: auto-stage modified tracked files
    if all_flag:
        entries = read_index(repo_root)
        head_sha = get_head_sha(repo_root)

        if head_sha:
            from objects import read_commit as rc, read_tree

            def get_committed_files(tree_sha: str, prefix: str = '') -> dict:
                files = {}
                try:
                    tree_entries = read_tree(repo_root, tree_sha)
                    for entry in tree_entries:
                        path = f"{prefix}{entry.name}" if prefix else entry.name
                        if entry.mode == '40000':
                            files.update(get_committed_files(entry.sha, path + '/'))
                        else:
                            files[path] = entry.sha
                except:
                    pass
                return files

            commit = rc(repo_root, head_sha)
            committed = get_committed_files(commit['tree'])

            from objects import create_blob

            for path, old_sha in committed.items():
                full_path = repo_root / path
                if full_path.exists():
                    content = full_path.read_bytes()
                    stat = full_path.stat()
                    mode = '100755' if stat.st_mode & 0o111 else '100644'
                    new_sha = create_blob(repo_root, content)
                    if new_sha != old_sha:
                        add_to_index(repo_root, path, new_sha, mode, stat)

    # Get index entries
    entries = read_index(repo_root)

    if not entries:
        print("error: nothing to commit", file=sys.stderr)
        return 1

    # Get parent commit
    head_sha = get_head_sha(repo_root)
    parents = []

    if amend_flag and head_sha:
        # Amend: use parent of current HEAD
        commit = read_commit(repo_root, head_sha)
        parents = commit['parents']
        if message is None:
            message = commit['message']
    elif head_sha:
        parents = [head_sha]

    # Check if there are any changes
    if head_sha and not amend_flag:
        from objects import read_tree

        def entries_match_tree(tree_sha: str, entries: list) -> bool:
            """Check if index entries match the tree exactly."""
            def get_tree_entries(sha: str, prefix: str = '') -> dict:
                result = {}
                try:
                    tree_entries = read_tree(repo_root, sha)
                    for entry in tree_entries:
                        path = f"{prefix}{entry.name}" if prefix else entry.name
                        if entry.mode == '40000':
                            result.update(get_tree_entries(entry.sha, path + '/'))
                        else:
                            result[path] = (entry.sha, entry.mode)
                except:
                    pass
                return result

            tree_files = get_tree_entries(tree_sha)
            index_files = {e['path']: (e['sha'], e['mode']) for e in entries}

            return tree_files == index_files

        commit = read_commit(repo_root, head_sha)
        if entries_match_tree(commit['tree'], entries):
            print("error: nothing to commit, working tree clean", file=sys.stderr)
            return 1

    # Build tree from index
    tree_sha = build_tree_from_index(repo_root, entries)

    # Create commit
    commit_sha = create_commit(repo_root, tree_sha, message, parents)

    # Update HEAD
    update_head_for_commit(repo_root, commit_sha)

    # Output commit SHA
    print(f"[{commit_sha[:7]}] {message.split(chr(10))[0]}")

    return 0
